### PR TITLE
CopyDataObjectFilter

### PR DESCRIPTION
**Pull request by jmarquisbq**
- CopyDataGroupFilter refactored and updated to be able to copy any DataObject instead of just groups
- Filter can now copy multiple DataObjects at once instead of just a single object
    - Added a MultiPathSelectionParameter for selecting multiple DataObjects (of any type) at once 

## Naming Conventions

Naming of variables should descriptive where needed. Loop Control Variables can use `i` if warranted. Most of these conventions are enforced through the clang-tidy and clang-format configuration files.

- [x] Class and Structs are `UpperCamelCase`
- [x] Class private member variables are `m_UpperCamelCase`
- [x] Class methods are `lowerCamelCase`
- [x] Method arguments are `lowerCamelCase`
- [x] Normal variables are `lowerCamelCase`
- [x] Constants are `k_UpperCamelCase`
- [x] Global statics are `s_UpperCamelCase`
- [x] Free Functions are `UpperCamelCase`
- [x] Macros are `ALL_UPPER_SNAKE_CASE`
- [x] Unit test will test data output from the filter.
- [x] Reused strings should be constants in an anonymous namespace
- [x] If parallelization is used, proper use of the abstracted complex classes are used. Using TBB specifically in a filter should be frowned upon unless for a really good reason.

## Filter Checklist

- [x] Parameters should be generally broken down into "Input Parameters", "Required Data Objects", "Created Data Objects". There can be exceptions to this.
- [x] ChoicesParameter selections should be an enumeration defined in the filer header
- [x] Documentation copied from SIMPL Repo and updated (if necessary)
- [x] Parameter argument variables are k_CamelCase_Key
- [x] Parameter argument strings are lower_snake_case
```
static inline constexpr StringLiteral k_AlignmentType_Key = "alignment_type";
```

## Unit Testing
- [x] 1 Unit test to instantiate the filter with the default arguments.
- [x] 1 Unit test to test output from the filter against known exemplar set of data
- [x] 1 Unit test to test invalid input code paths that are specific to a filter. Don't test that a DataPath does not exist since that test is already performed as part of the SelectDataArrayAction.

## Code Cleanup
- [x] No commented out code (rare exceptions to this is allowed..)
- [x] Filters should have both the Filter class and Algorithm class for anything beyond trivial needs
- [x] No API changes were made (or the changes have been approved)
- [x] No major design changes were made (or the changes have been approved)
- [x] Added test (or behavior not changed)
- [x] Updated API documentation (or API not changed)
- [x] Added license to new files (if any)
- [x] Added Python wrapping to new files (if any) as necessary
- [x] Added example pipelines that use the filter
- [x] Classes and methods are properly documented.

### CODE REVIEW

File: src/Plugins/ComplexCore/src/ComplexCore/Filters/CopyDataObjectFilter.cpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+#include "complex/Parameters/MultiPathSelectionParameter.hpp"
+#include "complex/Parameters/StringParameter.hpp"
+#include "complex/Utilities/DataGroupUtilities.hpp"
+#include "complex/Utilities/StringUtilities.hpp"
+
+namespace complex
+{
+namespace
+{
+} // namespace
```
Comment (nyoungbq) [inline src/Plugins/ComplexCore/src/ComplexCore/Filters/CopyDataObjectFilter.cpp]:
```suggestion
```
This empty anonymous namespace can be removed.

File: src/Plugins/ComplexCore/test/CopyDataObjectTest.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+#include "ComplexCore/Filters/CopyDataObjectFilter.hpp"
+
+#include "ComplexCore/ComplexCore_test_dirs.hpp"
+
+#include "complex/Parameters/MultiPathSelectionParameter.hpp"
+#include "complex/UnitTest/UnitTestCommon.hpp"
+
+using namespace complex;
+
+TEST_CASE("ComplexCore::CopyDataObjectFilter(Instantiate)", "[ComplexCore][CopyDataObjectFilter]")
```
Comment (nyoungbq) [inline src/Plugins/ComplexCore/test/CopyDataObjectTest.cpp]:
```suggestion
TEST_CASE("ComplexCore::CopyDataObjectFilter(Valid Execution)", "[ComplexCore][CopyDataObjectFilter]")
```

File: src/complex/Parameters/MultiPathSelectionParameter.hpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+#include "complex/DataStructure/IArray.hpp"
+#include "complex/Filter/MutableDataParameter.hpp"
+#include "complex/Filter/ParameterTraits.hpp"
+
+namespace complex
+{
+/**
+ * @brief This Filter Parameter represents multiple DataArrays as input.
```
Comment (nyoungbq) [inline src/complex/Parameters/MultiPathSelectionParameter.hpp]:
```suggestion
 * @brief This Filter Parameter represents multiple DataObjects as input.
```

File: src/complex/Parameters/MultiPathSelectionParameter.hpp [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  UniquePointer clone() const override;
+
+  /**
+   * @brief Returns the user defined default value.
+   * @return
+   */
+  std::any defaultValue() const override;
+
+  /**
+   * @brief
```
Comment (nyoungbq) [inline src/complex/Parameters/MultiPathSelectionParameter.hpp]:
```suggestion
   * @brief Returns the default path.
```

File: src/complex/Parameters/MultiPathSelectionParameter.hpp [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  /**
+   * @brief Validates the given value against the given DataStructure. Returns warnings/errors.
+   * @param dataStructure The active DataStructure to use during validation
+   * @param value The value to validate
+   * @return
+   */
+  Result<> validate(const DataStructure& dataStructure, const std::any& value) const override;
+
+  /**
+   * @brief Validates the given value against the given DataStructure. Returns warnings/errors.
```
Comment (nyoungbq) [inline src/complex/Parameters/MultiPathSelectionParameter.hpp]:
```suggestion
   * @brief Validates the given paths against the given DataStructure. Returns warnings/errors.
```

File: src/complex/Parameters/MultiPathSelectionParameter.hpp [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+   * @param value The value to validate
+   * @return
+   */
+  Result<> validatePaths(const DataStructure& dataStructure, const ValueType& value) const;
+
+  /**
+   * @brief Takes the value and a mutable DataStructure and attempts store the actual derived DataObject in the std::any.
+   * Returns any warnings/errors.
+   * @param dataStructure
+   * @param value
```
Comment (nyoungbq) [inline src/complex/Parameters/MultiPathSelectionParameter.hpp]:
```suggestion
   * @param dataStructure The active DataStructure to use during resolution
   * @param value The value to resolve
```

File: src/complex/Parameters/MultiPathSelectionParameter.cpp [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+Result<> MultiPathSelectionParameter::validatePaths(const DataStructure& dataStructure, const ValueType& value) const
+{
+  const std::string prefix = fmt::format("FilterParameter '{}' Validation Error: ", humanName());
+
+  std::vector<Error> errors;
+  for(usize i = 0; i < value.size(); i++)
+  {
+    const auto& path = value.at(i);
```
Comment (nyoungbq) [inline src/complex/Parameters/MultiPathSelectionParameter.cpp]: Did you mean to do a bounds check with the `.at(i)` ?
- If not switch to the bracket `[i]` operator for speed optimization.
- If you did I would add a comment to mention the bounds check is purposeful

File: src/complex/Parameters/MultiPathSelectionParameter.cpp [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+  return {typeid(ValueType)};
+}
+
+nlohmann::json MultiPathSelectionParameter::toJson(const std::any& value) const
+{
+  const auto& paths = GetAnyRef<ValueType>(value);
+  nlohmann::json json = nlohmann::json::array();
+  for(const auto& path : paths)
+  {
+    json.push_back(path.toString());
```
Comment (nyoungbq) [inline src/complex/Parameters/MultiPathSelectionParameter.cpp]:
```suggestion
    json.push_back(std::move(path.toString()));
```
I assume the string being returned is being assembled in the function and returned as an rvalue thus std::move ensures its an in-place move-construction preventing loss of efficiency

**Review by nyoungbq — COMMENTED** [review]
Might want to make sure the documentation changes I offered up are correct. Also I am not sure if it was already done, because I don't see it in the files changed, but make sure the documentation for this filter is updated/brought in

File: src/complex/Parameters/MultiPathSelectionParameter.cpp (hunk 7 above)
Comment (jmarquisbq) [inline src/complex/Parameters/MultiPathSelectionParameter.cpp]:
```suggestion
    const auto& path = value[i];
```

**Review by nyoungbq — APPROVED** [review]

**Pull request merged**